src/io/sb3/interfaces.ts

```diff
@@ -0,0 +1,502 @@
+import { OpCode } from "../../OpCode";
+import { KnownBlock, ProcedureBlock } from "../../Block";
+import { TextToSpeechLanguage } from "../../Project";
+import * as BlockInput from "../../BlockInput";
+
+// Note: This schema is designed to match the definitions in
+// https://github.com/LLK/scratch-parser/blob/master/lib/sb3_definitions.json
+
+// Values storable in variables and lists.
+export type ScalarValue = string | number | boolean;
+
+// 32-length hex string - the MD5 of the asset.
+// Does not include the asset's file extension.
+export type AssetId = string;
+
+// [name, value, cloud]
+// For example: ["Highscore", 3000, true]
+// Note: Scratch's server prevents uploading non-number values to the cloud
+// variable server, but this restriction is not enforced in the sb3 schema.
+export type Variable = [string, ScalarValue] | [string, ScalarValue, true];
+
+// [name, contents]
+// For example: ["My List", [1, 2, true, "banana"]]
+export type List = [string, ScalarValue[]];
+
+export interface Costume {
+  assetId: AssetId;
+  dataFormat: "png" | "svg" | "jpeg" | "jpg" | "bmp" | "gif";
+  name: string;
+
+  md5ext?: string;
+
+  bitmapResolution?: number;
+  rotationCenterX?: number;
+  rotationCenterY?: number;
+}
+
+export interface Sound {
+  assetId: AssetId;
+  dataFormat: "wav" | "wave" | "mp3";
+  name: string;
+
+  md5ext?: string;
+
+  rate?: number;
+  sampleCount?: number;
+}
+
+export interface Block {
+  opcode: OpCode;
+
+  next: string | null;
+  parent: string | null;
+
+  inputs: {
+    [key: string]: BlockInput;
+  };
+  fields: {
+    [key: string]: BlockField;
+  };
+
+  shadow: boolean;
+  topLevel: boolean;
+
+  x?: number;
+  y?: number;
+
+  mutation?: {
+    tagName: string;
+    children: any[]; // TODO: What is this?
```

Comment:
I think `mutation` is a JSON version of an XML node, and so `children` is an array of other XML nodes.

Reply:
I want to make the mutation attribute type better in any case - allowing for the specific mutations which Scratch supports. Right now it's a bit of a mish-mash of properties that don't matter as much as they should, typing-wise.

Reply:
I couldn't figure out how to make a clean type format for this, so I just boiled it to `{[attribute: string]: any}`, and have documented what it represents.